Add fast path to JSStringJoiner::join() when there is only one string in the vector

cdumez opened the pull request:
#### 47cdb9c389b608104995423112a03d532ceea386
<pre>
Add fast path to JSStringJoiner::join() when there is only one string in the vector
<a href="https://bugs.webkit.org/show_bug.cgi?id=254463">https://bugs.webkit.org/show_bug.cgi?id=254463</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::JSStringJoiner::joinSlow):
(JSC::JSStringJoiner::join): Deleted.
* Source/JavaScriptCore/runtime/JSStringJoiner.h:
(JSC::JSStringJoiner::join):

Canonical link: <a href="https://commits.webkit.org/262129@main">https://commits.webkit.org/262129@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96b098e93685855197819a87c4b2fcf68f4ae346

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/674 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/702 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/828 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/573 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/759 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/812 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/662 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/624 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/621 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/805 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/607 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/641 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/572 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/586 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/629 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/647 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/618 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/605 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/678 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/571 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/622 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/134 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/628 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/679 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/69 "Built successfully and passed tests") | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/128 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/JSStringJoiner.h:
```diff
@@ -77,6 +78,13 @@ inline JSStringJoiner::JSStringJoiner(JSGlobalObject* globalObject, LChar separa
+inline JSValue JSStringJoiner::join(JSGlobalObject* globalObject)
+{
+    if (m_strings.size() == 1)
+        return jsString(globalObject->vm(), m_strings[0].view.toString());
```
Why doesn't this use the underlying string if present? Seems that doing that could avoid allocating a new `StringImpl` in some cases. Maybe those aren't common enough.

cdumez replied:
I assumed the string view might point to only part of the underlying string? I am not very familiar with this code.

darinadler replied:
That’s correct. But:

1. if the lengths are equal, the underlying string can be used as is — I think this is a common case and the optimization of not allocating any memory for a `StringImpl` would be significant
2. if the string view’s length is smaller than the underlying string, we _could_ use `substringSharingImpl` to create a new `StringImpl` that is small and shares the characters, but that may be a more difficult tradeoff since we could keep a large string alive, and may not be as common a case

I was really thinking of (1) above, and if it’s common enough it might lead to a bigger performance win.

cdumez replied:
Thanks for the clarification. I'll see if we can get a bigger win.

cdumez replied:
You are right that (1) is the common case. 99.1% of the cases in Speedometer where m_strings has a length of 1 in fact. I'll do some A/B testing on the bots.

cdumez replied:
Hmm, I tried the following patch (https://pastebin.com/Nwacq8mF) but A/B testing says this is no longer a progression. Weird.

cdumez committed:
Add fast path to JSStringJoiner::join() when there is only one string in the vector
https://bugs.webkit.org/show_bug.cgi?id=254463

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::JSStringJoiner::joinSlow):
(JSC::JSStringJoiner::join): Deleted.
* Source/JavaScriptCore/runtime/JSStringJoiner.h:
(JSC::JSStringJoiner::join):

Canonical link: https://commits.webkit.org/262129@main
